File: SharedKit/Sources/SharedKit/Utilities/InputValidation.swift
import Foundation

/// Input validation utilities for task and orb creation
public enum InputValidation {
    
    // MARK: - Validation Errors
    
    public enum ValidationError: LocalizedError {
        case emptyTitle
        case titleTooLong(maxLength: Int)
        case titleContainsInvalidCharacters
        case emptyOrbName
        case orbNameTooLong(maxLength: Int)
        case invalidPriority(validRange: ClosedRange<Int>)
        case invalidStatus(validRange: ClosedRange<Int16>)
        case deadlineInPast
        case notesTooLong(maxLength: Int)
        
        public var errorDescription: String? {
            switch self {
            case .emptyTitle:
                return "Task title cannot be empty"
            case .titleTooLong(let maxLength):
                return "Task title cannot exceed \(maxLength) characters"
            case .titleContainsInvalidCharacters:
                return "Task title contains invalid characters"
            case .emptyOrbName:
                return "Project name cannot be empty"
            case .orbNameTooLong(let maxLength):
                return "Project name cannot exceed \(maxLength) characters"
            case .invalidPriority(let range):
                return "Priority must be between \(range.lowerBound) and \(range.upperBound)"
            case .invalidStatus(let range):
                return "Status must be between \(range.lowerBound) and \(range.upperBound)"
            case .deadlineInPast:
                return "Deadline cannot be in the past"
            case .notesTooLong(let maxLength):
                return "Notes cannot exceed \(maxLength) characters"
            }
        }
        
        public var recoverySuggestion: String? {
            switch self {
            case .emptyTitle:
                return "Please enter a title for your task"
            case .titleTooLong:
                return "Try shortening the title"
            case .titleContainsInvalidCharacters:
                return "Remove any special characters from the title"
            case .emptyOrbName:
                return "Please enter a name for your project"
            case .orbNameTooLong:
                return "Try shortening the project name"
            case .invalidPriority(let range):
                return "Set priority to a value between \(range.lowerBound) and \(range.upperBound)"
            case .invalidStatus(let range):
                return "Set status to a valid value"
            case .deadlineInPast:
                return "Choose a deadline in the future"
            case .notesTooLong:
                return "Try shortening the notes"
            }
        }
    }
    
    // MARK: - Configuration
    
    public struct Config {
        public let maxTitleLength: Int
        public let maxOrbNameLength: Int
        public let maxNotesLength: Int
        public let validPriorityRange: ClosedRange<Int>
        public let validStatusRange: ClosedRange<Int16>
        public let allowPastDeadlines: Bool
        
        public static let `default` = Config(
            maxTitleLength: 500,
            maxOrbNameLength: 100,
            maxNotesLength: 10000,
            validPriorityRange: 1...5,
            validStatusRange: 1...3,
            allowPastDeadlines: false
        )
        
        public init(
            maxTitleLength: Int = 500,
            maxOrbNameLength: Int = 100,
            maxNotesLength: Int = 10000,
            validPriorityRange: ClosedRange<Int> = 1...5,
            validStatusRange: ClosedRange<Int16> = 1...3,
            allowPastDeadlines: Bool = false
        ) {
            self.maxTitleLength = maxTitleLength
            self.maxOrbNameLength = maxOrbNameLength
            self.maxNotesLength = maxNotesLength
            self.validPriorityRange = validPriorityRange
            self.validStatusRange = validStatusRange
            self.allowPastDeadlines = allowPastDeadlines
        }
    }
    
    // MARK: - Task Validation
    
    /// Validates task input and returns sanitized title or throws validation error
    public static func validateTaskTitle(_ title: String, config: Config = .default) throws -> String {
        let trimmed = title.trimmingCharacters(in: .whitespacesAndNewlines)
        
        guard !trimmed.isEmpty else {
            throw ValidationError.emptyTitle
        }
        
        guard trimmed.count <= config.maxTitleLength else {
            throw ValidationError.titleTooLong(maxLength: config.maxTitleLength)
        }
        
        // Check for control characters (but allow unicode)
        let invalidCharacters = CharacterSet.controlCharacters.subtracting(.whitespaces)
        guard trimmed.unicodeScalars.allSatisfy({ !invalidCharacters.contains($0) }) else {
            throw ValidationError.titleContainsInvalidCharacters
        }
        
        return trimmed
    }
    
    /// Validates task priority
    public static func validatePriority(_ priority: Int, config: Config = .default) throws -> Int {
        guard config.validPriorityRange.contains(priority) else {
            throw ValidationError.invalidPriority(validRange: config.validPriorityRange)
        }
        return priority
    }
    
    /// Validates task status
    public static func validateStatus(_ status: Int16, config: Config = .default) throws -> Int16 {
        guard config.validStatusRange.contains(status) else {
            throw ValidationError.invalidStatus(validRange: config.validStatusRange)
        }
        return status
    }
    
    /// Validates deadline
    public static func validateDeadline(_ deadline: Date?, config: Config = .default) throws -> Date? {
        guard let deadline = deadline else { return nil }
        
        if !config.allowPastDeadlines && deadline < Date() {
            // Allow some tolerance (5 minutes) for recently set deadlines
            let tolerance: TimeInterval = 5 * 60
            if deadline.addingTimeInterval(tolerance) < Date() {
                throw ValidationError.deadlineInPast
            }
        }
        
        return deadline
    }
    
    /// Validates notes/details
    public static func validateNotes(_ notes: String?, config: Config = .default) throws -> String {
        guard let notes = notes else { return "" }
        let trimmed = notes.trimmingCharacters(in: .whitespacesAndNewlines)
        
        guard trimmed.count <= config.maxNotesLength else {
            throw ValidationError.notesTooLong(maxLength: config.maxNotesLength)
        }
        
        return trimmed
    }
    
    // MARK: - Orb Validation
    
    /// Validates orb name and returns sanitized name or throws validation error
    public static func validateOrbName(_ name: String, config: Config = .default) throws -> String {
        let trimmed = name.trimmingCharacters(in: .whitespacesAndNewlines)
        
        guard !trimmed.isEmpty else {
            throw ValidationError.emptyOrbName
        }
        
        guard trimmed.count <= config.maxOrbNameLength else {
            throw ValidationError.orbNameTooLong(maxLength: config.maxOrbNameLength)
        }
        
        return trimmed
    }
    
    // MARK: - Convenience Methods
    
    /// Sanitizes a string by trimming whitespace and limiting length
    public static func sanitize(_ input: String, maxLength: Int) -> String {
        let trimmed = input.trimmingCharacters(in: .whitespacesAndNewlines)
        if trimmed.count <= maxLength {
            return trimmed
        }
        return String(trimmed.prefix(maxLength))
    }
    
    /// Checks if a string is a valid non-empty input
    public static func isValidInput(_ input: String) -> Bool {
        return !input.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
    }
}

File: NotchToDo-Android/app/src/main/java/com/notchtodo/util/InputValidation.kt
package com.notchtodo.util

import java.util.Date

/**
 * Input validation utilities for task and orb creation.
 */
object InputValidation {
    
    // MARK: - Configuration
    
    data class Config(
        val maxTitleLength: Int = 500,
        val maxOrbNameLength: Int = 100,
        val maxNotesLength: Int = 10000,
        val validPriorityRange: IntRange = 1..5,
        val validStatusRange: IntRange = 1..3,
        val allowPastDeadlines: Boolean = false
    ) {
        companion object {
            val DEFAULT = Config()
        }
    }
    
    // MARK: - Validation Results
    
    sealed class ValidationResult<out T> {
        data class Success<T>(val value: T) : ValidationResult<T>()
        data class Error(val error: ValidationError) : ValidationResult<Nothing>()
        
        fun getOrNull(): T? = (this as? Success)?.value
        fun getOrThrow(): T = when (this) {
            is Success -> value
            is Error -> throw IllegalArgumentException(error.message)
        }
    }
    
    enum class ValidationError(val message: String, val suggestion: String) {
        EMPTY_TITLE(
            "Task title cannot be empty",
            "Please enter a title for your task"
        ),
        TITLE_TOO_LONG(
            "Task title is too long",
            "Try shortening the title"
        ),
        TITLE_INVALID_CHARACTERS(
            "Task title contains invalid characters",
            "Remove any special characters from the title"
        ),
        EMPTY_ORB_NAME(
            "Project name cannot be empty",
            "Please enter a name for your project"
        ),
        ORB_NAME_TOO_LONG(
            "Project name is too long",
            "Try shortening the project name"
        ),
        INVALID_PRIORITY(
            "Priority value is invalid",
            "Set priority to a value between 1 and 5"
        ),
        INVALID_STATUS(
            "Status value is invalid",
            "Set status to a valid value"
        ),
        DEADLINE_IN_PAST(
            "Deadline cannot be in the past",
            "Choose a deadline in the future"
        ),
        NOTES_TOO_LONG(
            "Notes are too long",
            "Try shortening the notes"
        )
    }
    
    // MARK: - Task Validation
    
    /**
     * Validates task title and returns sanitized title or error.
     */
    fun validateTaskTitle(
        title: String,
        config: Config = Config.DEFAULT
    ): ValidationResult<String> {
        val trimmed = title.trim()
        
        if (trimmed.isEmpty()) {
            return ValidationResult.Error(ValidationError.EMPTY_TITLE)
        }
        
        if (trimmed.length > config.maxTitleLength) {
            return ValidationResult.Error(ValidationError.TITLE_TOO_LONG)
        }
        
        // Check for control characters (but allow unicode)
        if (trimmed.any { it.isISOControl() && !it.isWhitespace() }) {
            return ValidationResult.Error(ValidationError.TITLE_INVALID_CHARACTERS)
        }
        
        return ValidationResult.Success(trimmed)
    }
    
    /**
     * Validates task priority.
     */
    fun validatePriority(
        priority: Int,
        config: Config = Config.DEFAULT
    ): ValidationResult<Int> {
        if (priority !in config.validPriorityRange) {
            return ValidationResult.Error(ValidationError.INVALID_PRIORITY)
        }
        return ValidationResult.Success(priority)
    }
    
    /**
     * Validates task status.
     */
    fun validateStatus(
        status: Int,
        config: Config = Config.DEFAULT
    ): ValidationResult<Int> {
        if (status !in config.validStatusRange) {
            return ValidationResult.Error(ValidationError.INVALID_STATUS)
        }
        return ValidationResult.Success(status)
    }
    
    /**
     * Validates deadline.
     */
    fun validateDeadline(
        deadline: Date?,
        config: Config = Config.DEFAULT
    ): ValidationResult<Date?> {
        if (deadline == null) return ValidationResult.Success(null)
        
        if (!config.allowPastDeadlines) {
            // Allow some tolerance (5 minutes) for recently set deadlines
            val tolerance = 5 * 60 * 1000L
            if (deadline.time + tolerance < System.currentTimeMillis()) {
                return ValidationResult.Error(ValidationError.DEADLINE_IN_PAST)
            }
        }
        
        return ValidationResult.Success(deadline)
    }
    
    /**
     * Validates notes/details.
     */
    fun validateNotes(
        notes: String?,
        config: Config = Config.DEFAULT
    ): ValidationResult<String> {
        if (notes == null) return ValidationResult.Success("")
        
        val trimmed = notes.trim()
        if (trimmed.length > config.maxNotesLength) {
            return ValidationResult.Error(ValidationError.NOTES_TOO_LONG)
        }
        
        return ValidationResult.Success(trimmed)
    }
    
    // MARK: - Orb Validation
    
    /**
     * Validates orb name and returns sanitized name or error.
     */
    fun validateOrbName(
        name: String,
        config: Config = Config.DEFAULT
    ): ValidationResult<String> {
        val trimmed = name.trim()
        
        if (trimmed.isEmpty()) {
            return ValidationResult.Error(ValidationError.EMPTY_ORB_NAME)
        }
        
        if (trimmed.length > config.maxOrbNameLength) {
            return ValidationResult.Error(ValidationError.ORB_NAME_TOO_LONG)
        }
        
        return ValidationResult.Success(trimmed)
    }
    
    // MARK: - Convenience Methods
    
    /**
     * Sanitizes a string by trimming whitespace and limiting length.
     */
    fun sanitize(input: String, maxLength: Int): String {
        val trimmed = input.trim()
        return if (trimmed.length <= maxLength) trimmed else trimmed.take(maxLength)
    }
    
    /**
     * Checks if a string is a valid non-empty input.
     */
    fun isValidInput(input: String): Boolean {
        return input.trim().isNotEmpty()
    }
}
